test/cloudpassage_lib/scans_test.clj

```
@@ -163,27 +179,114 @@
+                                  formatted-output-str (output-fn data)]
+                              (swap! log conj formatted-output-str)))]
+      (with-redefs [scans/get-page!
+                    (fn [client-id client-secret url]
+                      (let [{:keys [page-num path]}
+                            (parse-fake-request client-id client-secret url)]
+                        (is (= "/v1/servers" path))
+                        (case page-num
+                          1 {:servers [{:id "server-id-1"} {:id "server-id-2"}]
+                             :pagination
+                             {:next (str @#'scans/base-servers-url "?page=2")}}
+                          2 :cloudpassage-lib.core/fetch-error)))]
+        (use-atom-log-appender! log log-appender-fn)
+        (let [server-stream (scans/list-servers! "lvh" "hunter2")
+              id-list (map :id (ms/stream->seq server-stream))
+              expected-error (str "Error getting scans for url: "
+                                  "https://api.cloudpassage.com/v1/servers/?page=2")]
+          (is (str/includes? (first @log) expected-error)
+              "logs the page fetch error")
+          (is (= ["server-id-1" "server-id-2"] id-list))
+          (is (ms/closed? server-stream)))))))
+
+(deftest scan-each-server!-tests
+  (testing "Scan can handle an empty stream."
+    (let [input (ms/stream)]
+      (ms/close! input)
+      (let [scan-stream (scans/scan-each-server! "lvh" "hunter2" "svm" input)
+            scan-result (ms/stream->seq scan-stream)]
+        (is (empty? scan-result)))))
+  (testing "Returns an empty list when get-page returns ::fetch-error"
+    (with-redefs [scans/get-page!
+                  (fn [client-id client-secret url]
+                    (let [{:keys [path]}
+                          (parse-fake-request client-id client-secret url)]
+                      (is (= "/v1/servers/server-id-here/svm" path))
```

Wow, when I wrote that comment I just did a word substitution and didn't even mentally process that `parse-fake-request` is something we defined within the tests---I guess we are testing our tests here! Honestly, I don't think that's necessary, we can just return an error here imo. Will make it cleaner.